File: week_1/pset1/cash_more/credit.c
#include <stdio.h>
#include <cs50.h>
#define T 1000000000000   // number of digits including the one = number of digits of divisor
#define F 10000000000000
#define S1 1000000000000000
#define S2 100000000000000

int main(void)
{
    long ccnumber = get_long("Number: ");
    long cc = ccnumber;
    int count = 0;
    // indictes the length of the ccnumber
    while (cc != 0)
    {
        cc /= 10;
        count++;
    }
    /* printf("%i\n", count);
     printf("%ld\n", cc); */
    int sum1 = 0;
    int sum2 = 0;
    int t = 0;
    int s = 0;
    long number = ccnumber; //saving credit card number

    for (int n = 1; n <= count; n++) // doing the checksum
    {
        if (n % 2 == 1)
        {
            int i = ccnumber % 10; //taking the odds
            sum1 += i;    // sumation of odd digits (1,3,5,...)
        }
        else
        {
            int j = (ccnumber % 10) * 2;
            if (j >= 10 && j < 20) //check if the number contains of 2 digits or no
            {
                t += j % 10; //sums the last (second) digit
                s += j / 10; //sums the frist digit
            }
            else
            {
                sum2 += j;  //sums the even digits of one digit(2,4,6,8)
            }
        }
        ccnumber /= 10;  //replace the last place by the second-to-last digit
    }
    int lsum = sum1 + sum2 + s + t;

    // the card type
    if ((lsum % 10 == 0)  && (count == 13 || count == 16 || count == 15))
    {
        if ((count == 13  ||  count == 16) && (number / T == 4 || number / S1 == 4))
        {
            printf("VISA\n");
        }
        else if ((count == 15) && (number / F == 34 || number / F == 37))
        {
            printf("AMEX\n");
        }
        else if ((count == 16) && (number / S2 == 51 || number / S2 == 52 ||
                                   number / S2 == 53 || number / S2 == 54 || number / S2 == 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week_6/pset6/credit/credit.py
from cs50 import get_int

# some constants to get the first/second digit from the left
T = 1000000000000
F = 10000000000000
S1 = 1000000000000000
S2 = 100000000000000

# gets credit card number
while True:
    ccnumber = get_int("Number: ")
    if str(ccnumber).isdigit() == True:
        break
# saves credit card number
cc = ccnumber

# sum of odd places
sum1 = 0

# sum of even places
sum2 = 0

# the total sum
total = 0

# implements Luhn’s Algorithm
for i in range(1, len(str(ccnumber)) + 1):
    if i % 2 == 1:
        sum1 += cc % 10

    else:
        j = (cc % 10) * 2

        # sum of digits in numbers between 10:20
        if j in range(10, 20):
            sum2 += (j % 10) + int(j / 10)

        else:
            sum2 += j

    cc = int(cc / 10)

total = sum1 + sum2

# indicates credit type [VUSA, AMEX, MASTERCARD, or INVALID]
if (total % 10 == 0):
    if (len(str(ccnumber)) == 13 or len(str(ccnumber)) == 16) and (int(ccnumber / T) == 4 or int(ccnumber / S1) == 4):
        print("VISA")
    elif (len(str(ccnumber)) == 15) and (int(ccnumber / F) == 34 or int(ccnumber / F) == 37):
        print("AMEX")
    elif (len(str(ccnumber)) == 16) and (int(ccnumber / S2) == 51 or int(ccnumber / S2) == 52 or int(ccnumber / S2) == 53 or int(ccnumber / S2) == 54 or int(ccnumber / S2) == 55):
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")